src/main/java/apoc/coll/Coll.java

```diff
@@ -74,13 +75,62 @@ public double avg(@Name("numbers") List<Number> list) {
     @UserFunction
     @Description("apoc.coll.min([0.5,1,2.3])")
     public Object min(@Name("values") List<Object> list) {
-        return Collections.min((List)list);
+        return Collections.min((List)list, Coll::compareAsNumbers);
     }
 
     @UserFunction
     @Description("apoc.coll.max([0.5,1,2.3])")
     public Object max(@Name("values") List<Object> list) {
-        return Collections.max((List)list);
+        return Collections.max((List)list, Coll::compareAsNumbers);
+    }
+
+    private static int compareAsNumbers(Object a, Object b) {
+        if (isNaN(a) && isNaN(b)) {
+            return 0;
+        }
+        if (isNaN(a)) {
+            return 1;
+        }
+        if (isNaN(b)) {
+            return -1;
+        }
+
+        if (isPositiveInfinity(a) && isPositiveInfinity(b)) {
+            return 0;
+        }
+        if (isPositiveInfinity(a)) {
+            return 1;
+        }
+        if (isPositiveInfinity(b)) {
+            return -1;
+        }
+
+        if (isNegativeInfinity(a) && isNegativeInfinity(b)) {
+            return 0;
+        }
+        if (isNegativeInfinity(a)) {
+            return -1;
+        }
+        if (isNegativeInfinity(b)) {
+            return 1;
+        }
+
+        return (new BigDecimal((a.toString())).compareTo(new BigDecimal(b.toString())));
```

Comment:
I feel this is quite expensive? WDYT? Could we just cast them to Number and compare the double-values?

Reply:
I'll create a small performance tests to compare both approaches and check back.
Comparing based on the double value of the numbers might introduce some inaccuracy for larger numbers, but most likely this will be the fastest solution.

Maybe we could use the double-value for floating point numbers, and long-value for non-floating point numbers to avoid inaccuracy and BigDecimals as well.

Reply:
Sounds good
What I'm also wondering is if Double.nan and infinity has some builtin comparison logic?

Reply:
The Double.compare static method handles it. Tricky part will be to compare doubles to longs and handle these special cases as well.

Reply:
I created a small benchmark to compare the different approaches, you can check it out here: https://github.com/dodie/generic-number-comparator-benchmark


| Benchmark                          | Score (ops/s) - Comparing Integers only    | Score (ops/s) - Comparing Integers and Doubles |
| ---------------------------------- | -------------------------------- | ------------------------------------ |
| Original comparator                | 3172804.081                      | X                                    |
| [Long.compare](https://github.com/dodie/generic-number-comparator-benchmark/blob/master/src/main/java/hu/awm/test/comparison/LongValueBasedComparison.java)                       | 3480229.820                      | X                                    |
| [Double.compare](https://github.com/dodie/generic-number-comparator-benchmark/blob/master/src/main/java/hu/awm/test/comparison/DoubleValueBasedComparison.java)                     | 2756652.584                      | 2340389.827                          |
| [Double.compare / Long.compare](https://github.com/dodie/generic-number-comparator-benchmark/blob/master/src/main/java/hu/awm/test/comparison/LongAndDoubleBasedComparison.java)      | 2881121.264                      | 1899611.262                          |
| [BigDecimal.compare + Inf/NaN check](https://github.com/dodie/generic-number-comparator-benchmark/blob/master/src/main/java/hu/awm/test/comparison/BigDecimalBasedComparison.java) |  104337.061                      |   74441.372                          |


I put the original and the Long.compare based versions in the measurement as a baseline.

You were right, the BigDecimal based version is much slower than the other solutions.
The fastest comparator that can handle floating-point numbers is the Double.compare, but promoting longs to doubles might result in loss of precision. From the [JLS](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.1.2):


>A widening conversion of an int or a long value to float, or of a long value to double, may result in loss of precision - that is, the result may lose some of the least significant bits of the value. In this case, the resulting floating-point value will be a correctly rounded version of the integer value, using IEEE 754 round-to-nearest mode (§4.2.4).


[This](https://github.com/dodie/generic-number-comparator-benchmark/blob/master/src/main/java/hu/awm/test/comparison/LongAndDoubleBasedComparison.java) comparator is not much slower than ```Double.compare```, and it only uses the ```doubleValue``` if both numbers are floating-point values. Otherwise, it compares them as Longs.

I think it might be a good compromise between speed and accuracy.
What do you think, which version should we use?

Reply:
One more thing to consider is complexity: using Double.compare is much simpler than the alternatives.